File: output/gemma4-e2b/avatar/Python/Java/atcoder_ABC146_D.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.HashMap;
import java.util.Map;
import java.util.StringTokenizer;

public class atcoder_ABC146_D {

    // Helper class/structure to represent the edge key (min, max) for the map
    static class EdgeKey {
        int u, v;

        public EdgeKey(int a, int b) {
            this.u = Math.min(a, b);
            this.v = Math.max(a, b);
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            EdgeKey edgeKey = (EdgeKey) o;
            return u == edgeKey.u && v == edgeKey.v;
        }

        @Override
        public int hashCode() {
            return 31 * u + v;
        }
    }

    static int bfs(int v, int N, Deque<Integer>[] G, Map<EdgeKey, Integer> E) {
        int[] visited = new int[N];
        // visited[i] = 0 (unvisited), 1 (visited)
        
        Deque<Integer> queue = new ArrayDeque<>();
        Map<EdgeKey, Integer> node2color = new HashMap<>();
        
        int K = -1;
        
        queue.add(v);
        visited[v] = 1;
        
        // Initialize node2color for the starting node v. 
        // The Python code initializes node2color implicitly when it checks node2color[q].
        // We need a way to store the color assigned to nodes.
        // Since the Python code uses node2color[q] to determine the next color, 
        // we will use a map or array for node colors. Let's use an array for simplicity, 
        // assuming colors are assigned sequentially.
        int[] nodeColor = new int[N];
        nodeColor[v] = 0; // Start node color is 0 (or whatever the first color assignment implies)

        while (!queue.isEmpty()) {
            int q = queue.poll();
            int color = nodeColor[q]; // The color assigned to q
            
            int currentMaxColor = 0;

            for (int nex : G[q]) {
                if (visited[nex] == 0) {
                    visited[nex] = 1;
                    
                    // Calculate the next color based on the logic:
                    // color += 1
                    // if color == node2color[q]: color += 1
                    
                    int nextColor = color + 1;
                    if (nextColor == nodeColor[q]) {
                        nextColor += 1;
                    }
                    
                    nodeColor[nex] = nextColor;
                    
                    // Store edge weight/color
                    EdgeKey key = new EdgeKey(q, nex);
                    E.put(key, nextColor);
                    
                    queue.add(nex);
                    currentMaxColor = Math.max(currentMaxColor, nextColor);
                }
            }
            K = Math.max(K, currentMaxColor);
        }
        
        // The Python code seems to calculate K based on the maximum color assigned.
        // We return K.
        return K;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        // Read N
        int N = Integer.parseInt(br.readLine());
        
        // G: Adjacency list (Array of Deques)
        @SuppressWarnings("unchecked")
        Deque<Integer>[] G = new Deque[N];
        for (int i = 0; i < N; i++) {
            G[i] = new ArrayDeque<>();
        }
        
        // E: Edge map (using EdgeKey for pair representation)
        Map<EdgeKey, Integer> E = new HashMap<>();
        
        // Read N-1 edges
        for (int i = 0; i < N - 1; i++) {
            String line = br.readLine();
            if (line == null || line.isEmpty()) continue;
            
            StringTokenizer st = new StringTokenizer(line);
            
            // Input is 1-based, convert to 0-based
            int a = Integer.parseInt(st.nextToken()) - 1;
            int b = Integer.parseInt(st.nextToken()) - 1;
            
            G

File: output/gemma4-e2b/avatar/Java/Python/atcoder_ABC146_D.py
import sys
import io

# Set up input reading from stdin
input_data = sys.stdin.read().split()
if not input_data:
    # Handle case where input is empty
    exit()

# Use an iterator for reading input tokens
data_iter = iter(input_data)

def read_int():
    return int(next(data_iter))

# Read N
try:
    n = read_int()
except StopIteration:
    # Should not happen if input is well-formed
    exit()

# Initialize graph G (Adjacency List)
# G will store lists of Edge objects or tuples representing neighbors and edge IDs
g = [[] for _ in range(n)]

# Read N-1 edges and build the graph
for i in range(n - 1):
    try:
        # a and b are 1-based in input, convert to 0-based indices
        a = read_int() - 1
        b = read_int() - 1
    except StopIteration:
        # Ran out of input unexpectedly
        break

    # The Java code adds edges (i, b) and (i, a) where i is the edge index (0 to n-2)
    # This suggests the edge index i corresponds to the edge being added.
    # The edge IDs in the Java code seem to be 0 to n-2.
    
    # We store the edge information along with the destination node and the edge ID
    # Since the Java code uses Edge objects with an 'id', we need to track edge IDs explicitly.
    
    # Let's adjust the graph structure to store (neighbor, edge_id)
    # The Java code uses Edge(i, b) and Edge(i, a) where i is the edge index (0 to n-2).
    
    # We need to store the edge ID associated with the connection.
    # Since the DFS relies on ans[e.id], we must ensure e.id corresponds to the edge index.
    
    # Let's redefine G to store (neighbor, edge_index)
    
    # Re-reading the Java logic:
    # g.get(a).add(new Edge(i, b)); g.get(b).add(new Edge(i, a));
    # i here is the index of the edge being processed (0 to n-2).
    
    # We will store tuples (neighbor, edge_index) in the adjacency list.
    
    # Edge index for this edge is i
    g[a].append((b, i))
    g[b].append((a, i))


# Initialize ans array for n-1 edges
ans = [0] * (n - 1)

# DFS function implementation
# to: current node
# color: the 'color' parameter from the Java code (used for comparison)
# parents: parent node
def dfs(to, color, parents):
    # k tracks the assigned value (1-based)
    k = 1
    
    # Iterate over neighbors (neighbor, edge_id)
    for neighbor, edge_id in g[to]:
        if neighbor == parents:
            continue
        
        # Check the condition: if k == color, increment k
        if k == color:
            k += 1
            
        # Assign the result to ans[edge_id]
        # Note: The Java code assigns ans[e.id] = k. e.id is the edge index.
        ans[edge_id] = k
        
        # Recursive call
        dfs(neighbor, k, to)
        
        # Increment k after the recursive call returns (this corresponds to the Java k++)
        k += 1

# Start DFS from node 0, initial color/value is -1 (or whatever the Java code implies)
# The Java call was dfs(0, -1, -1)
dfs(0, -1, -1)

# Find max
max_val = 0
if ans:
    max_val = max(ans)

# Print results
print(max_val)
for result in ans:
    print(result)